proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,324 @@
+* Quick glance over `wal` or `wbl` directory tells if the migration to a next version happened or not.
+* We can extend the filename format in future segment version, cleanly e.g. another suffix for sharding.
+
+The mixed situation could look as follows:
+
+```
+wal/ 
+├── 00000000
+├── 00000001-v2
+├── 00000002-v2
+├── 00000003-v2
+├── 00000004
+├── checkpoint.00000001
+│   └── 00000000
+├── checkpoint.00000003
+│   └── 00000000-v2
+└── checkpoint.00000004
+    └── 00000000
+```
+
+We propose to document that behaviour in:
+
+* dedicated guide for users
+* flag help
+* tsdb/docs for devs
+
+Finally, we propose that all Prometheus releases will contain the content similar to the following one:
+
+| Data        | Supported | Writes |
+|-------------|-----------|--------|
+| WAL         | v1, v2    | v1     |
+| Block index | v2        | v2     |
+
+```
+The last revertable Prometheus version: v3.0.0 
+```
+
+See the [alternatives](#alternatives) for other ideas.
+
+### How: Why Flag?
+
+There are two reasons for the flag:
+
+* Allows users to get the new features sooner and skip the safety mechanism.
+* It simplifies the compatibility guarantees as the flag default mechanism guides users and devs in the rollout and revert procedures e.g:
+  * when we switch to writing v2, it's clearly visible in a flag default value. 
+  * when we at some point remove support of WAL v1, it's clear when it happens (`v1` flag value is removed).
+* It allows users to set Prometheus to write the old format if needed (de-risking further).
+* Gives devs quicker feedback, makes testing easier, and motivates further contributions.
+
+### How: Two-Fold Migration Strategy
+
+Given the following example:
+
+![twofold.png](../assets/2024-11-25_changing_wal_format/twofold.png)
+
+1. We release Prometheus X+1 version that supports both Y and Y+1 data but still writes Y.
+2. We release Prometheus X+2 version that supports both Y and Y+1 data, but now it writes new data as Y+1.
+
+While this example shows only one version where of forward compatibility (when Y and Y+1 are supported, but Y is still written), in practice there could be more "forward compatible releases" within this strategy.
+
+Technically we have to keep old format read (and write) support until the next major release.
+
+Pros:
+* Users have a durable rollout path back and forth.
+* Dev has clarity on how to develop "breaking revert" changes.
+* In theory, it allows revert to X from X+2, by going through X+1 and ensuring all data was migrated (eventually) to Y version. In practice however that eventuality is long, or hard to discover.
+
+Cons:
+* It can catch users by surprise.
+  * **Mitigation**: We also plan guide and documentation and support LTS release.
+* It takes time to rollout new changes if you want them fast.
+  * **Mitigation**: We plan to add a flag to opt-in sooner.
+
```

Comment:
Would it be viable to extend the WAL replay so that if there are unknown records in it, probably because you've rolled back to an older version, we keep take them out of the WAL file (or just keep them on disk if we use the `-vX` alternative) and keep them there until the user rolls forward again to a version that can consume them?